wavesrv/pkg/cmdrunner/cmdrunner.go

```diff
@@ -1609,48 +1635,59 @@ func CopyFileCommand(ctx context.Context, pk *scpacket.FeCommandPacketType) (sst
+	var cmd *sstore.CmdType
+	var update *sstore.ModelUpdate
+	outputPty := resolveBool(pk.Kwargs["outputpty"], false)
+	log.Printf("outputpty: %v", outputPty)
+	if !outputPty {
+		outputStr := fmt.Sprintf("Copying [%v]:%v to [%v]:%v\r\n", sourceRemoteId.DisplayName, sourceFullPath, destRemoteId.DisplayName, destFullPath)
+		termopts := sstore.TermOpts{Rows: shellutil.DefaultTermRows, Cols: shellutil.DefaultTermCols, FlexRows: true, MaxPtySize: remote.DefaultMaxPtySize}
+		cmd, err = makeDynCmd(ctx, "copy file", ids, pk.GetRawStr(), termopts)
+		writeStringToPty(ctx, cmd, outputStr, &outputPos, outputPty)
+		if err != nil {
+			// TODO tricky error since the command was a success, but we can't show the output
+			return nil, err
+		}
+		update, err := addLineForCmd(ctx, "/copy file", false, ids, cmd, "", nil)
+		if err != nil {
+			// TODO tricky error since the command was a success, but we can't show the output
+			return nil, err
+		}
+		update.Interactive = pk.Interactive
+		sstore.MainBus.SendScreenUpdate(cmd.ScreenId, update)
+	} else {
+		cmd, err = sstore.GetCmdByScreenId(ctx, pk.Kwargs["screenid"], pk.Kwargs["lineid"])
+		if err != nil {
+			return nil, err
+		}
 	}
-	update.Interactive = pk.Interactive
 	if destRemote != ConnectedRemote && destRemoteId != nil && !destRemoteId.RState.IsConnected() {
-		writeStringToPty(ctx, cmd, fmt.Sprintf("Attempting to autoconnect to remote %v\r\n", destRemote), &outputPos)
+		writeStringToPty(ctx, cmd, fmt.Sprintf("Attempting to autoconnect to remote %v\r\n", destRemote), &outputPos, outputPty)
 		err = destRemoteId.MShell.TryAutoConnect()
 		if err != nil {
-			writeStringToPty(ctx, cmd, fmt.Sprintf("Couldn't connect to remote %v\r\n", sourceRemote), &outputPos)
+			writeStringToPty(ctx, cmd, fmt.Sprintf("Couldn't connect to remote %v\r\n", sourceRemote), &outputPos, outputPty)
 		} else {
-			writeStringToPty(ctx, cmd, "Auto connect successful\r\n", &outputPos)
+			writeStringToPty(ctx, cmd, "Auto connect successful\r\n", &outputPos, outputPty)
 		}
 	}
 	if sourceRemote != LocalRemote && sourceRemoteId != nil && !sourceRemoteId.RState.IsConnected() {
-		writeStringToPty(ctx, cmd, fmt.Sprintf("Attempting to autoconnect to remote %v\r\n", sourceRemote), &outputPos)
+		writeStringToPty(ctx, cmd, fmt.Sprintf("Attempting to autoconnect to remote %v\r\n", sourceRemote), &outputPos, outputPty)
 		err = sourceRemoteId.MShell.TryAutoConnect()
 		if err != nil {
-			writeStringToPty(ctx, cmd, fmt.Sprintf("Couldn't connect to remote %v\r\n", sourceRemote), &outputPos)
+			writeStringToPty(ctx, cmd, fmt.Sprintf("Couldn't connect to remote %v\r\n", sourceRemote), &outputPos, outputPty)
 		} else {
-			writeStringToPty(ctx, cmd, "Auto connect successful\r\n", &outputPos)
+			writeStringToPty(ctx, cmd, "Auto connect successful\r\n", &outputPos, outputPty)
 		}
 	}
-	sstore.MainBus.SendScreenUpdate(cmd.ScreenId, update)
 	update = &sstore.ModelUpdate{}
 	if destRemote == LocalRemote && sourceRemote == LocalRemote {
```

Comment:
I'm thinking of refactoring this so that all copy files to "copyfileRemoteToRemote" and go through the mshell

This might be slower but it will remove a ton of boilerplate and make the code a lot more readable, and I think, since latency to the local wave shell is not currently a concern, it shouldn't impact speed